User view courses

GuiRokk opened the pull request:
Criando meios para um usuário visitante ver os cursos e planos disponíveis na plataforma
Adequando as rotas de Planos já existentes para o escopo de adminstrador

Commit:
WIP View lessons
Co-authored-by: Lucas Benevides <lBenevides@users.noreply.github.com>

Commit:
Update with subscriptions
Co-authored-by: Lucas Benevides <lBenevides@users.noreply.github.com>

Commit:
Add user restriction to view lessons
Co-authored-by: Lucas Benevides <lBenevides@users.noreply.github.com>

Commit:
Refactor subscription routes to admin scope

Co-authored-by: Lucas Benevides <lBenevides@users.noreply.github.com>

JorgeLAB commented on app/controllers/admins/courses_controller.rb:
```diff
@@ -6,4 +6,17 @@ def index
     @course =  Course.find(params[:id])
   end
+
+  def add
```
Talvez essa action não tenha um nome auto-explicativo, `admin_courses_add_path`, seria para adicionar onde ?  Pela lógica implementada talvez seja uma regra de negócio vinculado ao controller subscription, já que, a finalidade é acrescentar um curso à um plano tendo em vista que ele não o tenha.  
`@course = Course.find(params[:course_id])` acredito que fica explicíto que existe uma curso com id = course_id, já que, vamos ter de listar cursos possíveis para o admin cadastrar. Talvez o primeiro guard clause seja dispensável.

HenriqueMorato replied:
Acho que o que não ficou muito claro na minha visão é que esse add tá no courses_controller. Se ele tivesse como um `add_course` no subscriptions_controller talvez ficasse bem mais claro o que deveria acontecer só de ler.

JorgeLAB commented on app/controllers/admins/courses_controller.rb:
```diff
@@ -6,4 +6,17 @@ def index
+    return redirect_to [:admins, @subscription], alert: 'Curso não encontrado' unless @course
+
+    if @subscription.courses.include?(@course)
+      redirect_to [:admins, @subscription], alert: 'Curso já cadastrado nesta assinatura'
```
Aqui, se a adição fosse responsabilidade de uma rota subscription ( no controller Subscriptions ) seria apenas só dar um render na mesma view em caso de erro não redirecionaria para outra view.
Tipo:
https://api.rubyonrails.org/v6.1.3.1/classes/ActiveRecord/FinderMethods.html#method-i-exists-3F  => `exists?`
```
unless @subscription.course.exists?(@course)
  @subscription.course << @course # talvez possa abstrair em um método de Subcription, jogar no model
   redirect_to [:admin, @subscription] , notice: t('.success')
else
  flash[:notice] = ''Curso já cadastrado nesta assinatura" 
  render :add  # este add poderia ser apenas uma rota post nas rotas ao invés de ser uma view  
                      # igual ao caso do disable e enable no system-promotion que fizemos na campus 
end
```
É bem parecido com o que fizemos no system-promotion de habilitar e desabilitar, no caso seria `add_course` a um plano ou `remove_course` de um plano.

HenriqueMorato replied:
Aqui acho que o include? geraria a mesma query do exists?

https://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-include-3F

Ele fala um pouco sobre isso na doc

JorgeLAB commented on app/views/admins/subscriptions/index.html.erb:
```diff
@@ -0,0 +1,16 @@
+<hr>
+<% @subscriptions.each do |subscription| %>
+  <div id="<%= subscription.id %>">
+    <%= link_to subscription.name, admins_subscription_path(subscription) %><br>
```
Acho que alguns links com `admins` irão quebrar, já que, definimos um padrão de namespace como `admin`. o/

lBenevides replied:
Então, a gente ja puxou do main e veio como Admins o namespace

JorgeLAB commented on app/views/admins/subscriptions/show.html.erb:
```diff
@@ -0,0 +1,50 @@
+<% end %>
+
+
+<% if !@courses.nil? %>
```
```suggestion
<% if @courses.present? %>
```

JorgeLAB commented on config/routes.rb:
```diff
@@ -2,22 +2,30 @@
+    resources :subscriptions, only: [:index, :show, :new, :create] do
+      get 'search_course', on: :member
+      resources :courses, only: [] do
+        post 'add', on: :member
```
post o/ blz vi agora

JorgeLAB commented on spec/system/admins/subscription_spec.rb:
```diff
@@ -2,45 +2,52 @@
+    login_as admin, scope: :admin
+    visit admins_subscriptions_path
+
     expect(page).to have_text(subscription.name)
```
Talvez seja interessante a verificação da existência de mais de um plano presente na index de plano. Algo como:
```
subscriptions = Fabricate.times(3, :subscription) 
 ...
subscriptions.each do |subscription|
  expect(page).to have_text(subscription.description)
  expect(page).to have_text(other_subscription.name)
  expect(page).to have_text(other_subscription.description)
end
```
Foi até criado o other_subscription mas não foi verificado a existência na página o/

lBenevides committed:
Atualizando namespace de Admins para Admin e resolvendo conflitos

lBenevides committed:
visitante e usuario consegue visitar planos disponiveis

lBenevides committed:
Incluido i18 no Admin

Co-authered-by: Guilherme Roque <https://github.com/GuiRokk>

HenriqueMorato commented on app/views/admin/subscriptions/show.html.erb:
```diff
@@ -0,0 +1,51 @@
+  <% end %>
+
+<% if @courses.present? %>
+  <%= render partial: "course" %>
```
Essa partial tá gerando um select pra cada curso, não seria um select só para todos os cursos?

HenriqueMorato commented on app/controllers/admin/courses_controller.rb:
```diff
@@ -6,4 +6,17 @@ def index
   end
+
+  def add
+    @subscription = Subscription.find(params[:id])
+    @course = Course.find(params[:course_id])
+    return redirect_to [:admin, @subscription], alert: 'Curso não encontrado' unless @course
+
+    if @subscription.courses.include?(@course)
+      redirect_to [:admin, @subscription], alert: 'Curso já cadastrado nesta assinatura'
+    else
+      @subscription.courses << @course
+      redirect_to [:admin, @subscription], alert: 'Curso cadastrado com sucesso'
+    end
```
Eu acho que essa action tem lógica demais, como na view é um select você não precisa tratar tudo que chegar. Porque se a pessoa executar algo malicioso ela é a responsável.

Por exemplo:

```ruby
 @course = Course.find(params[:course_id])
return redirect_to [:admin, @subscription], alert: 'Curso não encontrado' unless @course
```

Essa segunda linha nunca vai acontecer porque se o find falhar sobe um erro e não existe um rescue então já daria erro 500 na tela. E não precisa tratar isso porque se a pessoa usar o select bonitinho ela não vai receber esse erro.

```ruby
 if @subscription.courses.include?(@course)
  redirect_to [:admin, @subscription], alert: 'Curso já cadastrado nesta assinatura'
```

Essa próxima você pode tratar com duas verificação uma é quando for preencher os cursos no select não mandar cursos que já estão nessa assinatura, algo como `@courses = Course.where.not(subscription: @subscription)`

A segunda seria adicionar no banco um index único entre essas duas tabelas:

`add_index :subscription_courses, [:couse_id, :subscription_id], unique: true`

Dessa forma um curso nunca pode estar duas vezes na mesma subscription (lembre-se de fazer testes de model)

Isso simplificaria essa action pra basicamente duas linhas e deixa seu controller com muito menos acoplamento

Commit:
Refactor PR Observations

Co-authored-by:
Henrique Morato <HenriqueMorato@users.noreply.github.com>
Lucas Benevides <lBenevides@users.noreply.github.com>

Commit:
Add migration unique course subscription id on ref

Co-authored-by: Lucas Benevides <lBenevides@users.noreply.github.com>

GuiRokk commented:
Adcionado as rotas, e controllers para um usuário de visitante:
 - Visualizar planos discponíveis na plataforma
 - Visualizar os cursos disponíveis na plataforma
 - View de comprar um curso

Adequação de rotas de escopo de admin para cadastrar novos planos no site.
I18n
Bootstrap básico das views

Commit:
Resolved Conflicts

Co-authored-by: Lucas Benevides <lBenevides@users.noreply.github.com>
